File: JAVA/RestoreIPAddresses.java
import java.util.ArrayList;

class Solution {
	ArrayList<String> ret = new ArrayList<String>();
    public ArrayList<String> restoreIpAddresses(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
    	ret.clear();
        if(s == null || s.length() <= 3 || s.length() > 12)
        	return ret;
        ArrayList<Integer> list = new ArrayList<Integer>();
        dfs(0, s.length(), s, list);
        return ret;
    }
	private void dfs(int index, int len, String s, ArrayList<Integer> list) {
		if(index == len)
		{
			if(list.size() != 4)
				return;
			StringBuilder sb = new StringBuilder();
			int i = 0;
			for(; i < 3; i++)
			{
				sb.append(String.valueOf(list.get(i))).append('.');
			}
			sb.append(String.valueOf(list.get(i)));
			ret.add(sb.toString());
			return;
		}
		for(int i = 1; i <= 3 && index+i <= len; i++)
		{
			String str = s.substring(index, index+i);
			if(i > 1 && str.charAt(0) == '0')
				continue;
			int cur = Integer.parseInt(str);
			if(cur > 255)
				continue;
			list.add(cur);
			dfs(index+i, len, s, list);
			list.remove(list.size()-1);
		}
	}
    
}
public class RestoreIPAddresses {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C++/RestoreIPAddresses.cpp
#include<iostream>
#include<stdlib.h>
#include<sstream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
	vector<string> ret;
	void dfs(int index, int size, string s, vector<string>& path)
	{
		if(index >= size)
		{
			if(path.size() == 4)
			{
				string str;
				for(int i = 0; i < 3; i++)
				{
					str.append(path[i]).push_back('.');
				}
				str.append(path[3]);
				ret.push_back(str);
			}
			return;
		}
		char tmp[4];
		stringstream sst;
		for(int len = 1; len <= 3 && index + len <= size; len++)
		{
			string cur = s.substr(index, len);
			if(len > 1 && cur[0] == '0')
				continue;
			int d = atoi(cur.c_str());
			if(d >= 0 && d <= 255)
			{
				//path.push_back(itoa(d,tmp,10));
				sst.str("");
				sst<<d;
				//cout<<d<<" "<<sst.str()<<"."<<endl;
				path.push_back(sst.str());
				dfs(index+len, size,s, path);
				path.pop_back();
			}
		}
	}
    vector<string> restoreIpAddresses(string s) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		ret.clear();
		vector<string> path;
		dfs(0, s.size(), s, path);
		return ret;
    }
};
int main()
{
	string s = "010010";
	Solution sol;
	sol.restoreIpAddresses(s);
	getchar();
}